extensions/amp-a4a/0.1/signature-verifier.js

```diff
@@ -85,3 +88,343 @@ export class ISignatureVerifier {
+   * endpoint URL as a query parameter and the keyset is re-fetched. Other kinds
+   * of failures, including network connectivity failures, are not retried.
+   *
+   * @param {string} signingServiceName
+   * @param {string} keypairId
+   * @param {!Uint8Array} signature
+   * @param {!ArrayBuffer} creative
+   * @param {function(string, !Object)} lifecycleCallback called for each AMP
+   *     lifecycle event triggered during verification
+   * @return {!Promise<!VerificationStatus>}
+   * @visibleForTesting
+   */
+  verifyCreativeAndSignature(
+      signingServiceName, keypairId, signature, creative, lifecycleCallback) {
+    if (!this.signers_) {
+      // Web Cryptography isn't available.
+      return Promise.resolve(VerificationStatus.UNVERIFIED);
+    }
+    const signer = this.signers_[signingServiceName];
+    return signer.promise.then(success => {
+      if (!success) {
+        // The public keyset couldn't be fetched and imported. Probably a
+        // network connectivity failure.
+        return VerificationStatus.UNVERIFIED;
+      }
+      const keyPromise = signer.keys[keypairId];
+      if (keyPromise === undefined) {
+        // We don't have this key, but maybe the cache is stale; try
+        // cachebusting.
+        signer.promise =
+            this.fetchAndAddKeys_(signer.keys, signingServiceName, keypairId)
+                .then(success => {
+                  if (signer.keys[keypairId] === undefined) {
+                    // We still don't have this key; make sure we never try
+                    // again.
+                    signer.keys[keypairId] = null;
+                  }
+                  return success;
+                });
+        // This "recursive" call can recurse at most once.
+        return this.verifyCreativeAndSignature(
+            signingServiceName, keypairId, signature, creative,
+            lifecycleCallback);
+      } else if (keyPromise === null) {
+        // We don't have this key and we already tried cachebusting.
+        return VerificationStatus.ERROR_KEY_NOT_FOUND;
+      } else {
+        return keyPromise.then(key => {
+          if (!key) {
+            // This particular public key couldn't be imported. Probably the
+            // signing service's fault.
+            return VerificationStatus.UNVERIFIED;
+          }
+          const crypto = Services.cryptoFor(this.win_);
+          const startTime = this.getNow_();
+          return crypto.verifyPkcs(key, signature, creative).then(
+              result => {
+                const endTime = this.getNow_();
+                if (result) {
+                  lifecycleCallback('signatureVerifySuccess', {
+                    'met.delta.AD_SLOT_ID': endTime - startTime,
+                    'signingServiceName.AD_SLOT_ID': signingServiceName,
+                  });
+                  return VerificationStatus.OK;
+                } else {
+                  return VerificationStatus.ERROR_SIGNATURE_MISMATCH;
```

Comment:
I'd prefer we send a CSI ping in both cases... certainly it would be of value so we could track how often client-validation fails but perhaps that's out of scope for this PR.  At last add a TODO.